File: src/SelectionSort.php
<?php

namespace Victor\GrokkingAlgorithmsExercises;

final class SelectionSort
{
    public function __construct(private array $items = []) {}

    /**
     * Searches the smallest index of a given array
     */
    private function searchSmallest(): int
    {
        $smallest = $this->items[0];
        $smallestIndex = 0;

        foreach ($this->items as $index => $item) {
            if ($item < $smallest) {
                $smallest = $item;
                $smallestIndex = $index;
            }
        }

        return $smallestIndex;
    }

    /**
     * Orders items by search the smallest index on each iteration, Big O execution time: O(n^2)
     */
    public function execute(): array
    {
        $orderedItems = [];

        foreach ($this->items as $_) {
            $smallestIndex = $this->searchSmallest();

            $orderedItems[] = array_splice($this->items, $smallestIndex, 1)[0];
        }

        return $orderedItems;
    }
}

File: lib/grokking_algorithms_exercises/selection_sort.rb
# frozen_string_literal: true

module GrokkingAlgorithmsExercises
  # Selection sort algorithm, Big O execution time: O(n^2)
  class SelectionSort
    def initialize(items = [])
      @items = items.dup
    end

    # Orders the items by taking the smallest one on each iteration
    def execute
      ordered_items = []

      ordered_items << @items.delete_at(search_smallest) until @items.empty?

      ordered_items
    end

    private

    # Searches the index of the smallest of the remaining items
    def search_smallest
      smallest = @items[0]
      smallest_index = 0

      @items.each_with_index do |item, index|
        if item < smallest
          smallest = item
          smallest_index = index
        end
      end

      smallest_index
    end
  end
end
